azurelinuxagent/common/persist_firewall_rules.py

```diff
@@ -0,0 +1,219 @@
+
+    def __verify_firewall_rules_enabled(self):
+        # Check if firewall-rules have already been enabled
+        # This function would also return False if the dest-ip is changed. So no need to check separately for that
+        try:
+            AddFirewallRules.check_firewalld_rule_applied(self._wait, self._dst_ip, self._uid)
+        except Exception as error:
+            logger.info(
+                "Check if Firewall rules already applied using firewalld.service failed: {0}".format(ustr(error)))
+            return False
+
+        return True
+
+    def _setup_permanent_firewalld_rules(self):
+        if self.__verify_firewall_rules_enabled():
+            logger.info("Firewall rules already set. No change needed.")
+            return
+
+        logger.info("Firewall rules not added yet, adding them now using firewalld.service")
+        # Add rules if not already set
+        AddFirewallRules.add_firewalld_rules(self._wait, self._dst_ip, self._uid)
+        logger.info("Successfully added the firewall commands using firewalld.service")
+
+    def __verify_network_setup_service_enabled(self):
+        # Check if the custom service has already been enabled
+        cmd = ["systemctl", "is-enabled", self._network_setup_service_name]
+        try:
+            return shellutil.run_command(cmd).rstrip() == "enabled"
+        except CommandError as error:
+            msg = "{0} not enabled. Command: {1}, ExitCode: {2}\nStdout: {3}\nStderr: {4}".format(
+                self._network_setup_service_name, ' '.join(cmd), error.returncode, error.stdout, error.stderr)
+        except Exception as error:
+            msg = "Ran into error, {0} not enabled. Error: {1}".format(self._network_setup_service_name, ustr(error))
+
+        logger.info(msg)
+        return False
+
+    def _setup_network_setup_service(self):
+        if self.__verify_network_setup_service_enabled():
+            logger.info("Service: {0} already enabled. No change needed.".format(self._network_setup_service_name))
+        else:
+            # Ensure waagent_network_setup.py is available in the desired spot
+            self.__set_network_setup_bin_file()
+            # Create unit file with default values
+            self.__set_service_unit_file()
+
+        # Even if service is enabled, we need to overwrite the drop-in file with the current IP incase it changed.
+        # This is to handle the case where WireIP can change midway on service restarts.
+        self.__set_drop_in_file()
+
+    def __set_network_setup_bin_file(self):
+        if os.path.exists(self._agent_network_setup_bin_file):
+            logger.verbose("{0} file exists in the expected place".format(self._agent_network_setup_bin_file))
+            return
+
+        logger.warn("Network file: {0} not available in the expected path. Copying it over".format(
+            self._agent_network_setup_bin_file))
+        try:
+            parent, _ = os.path.split(os.path.abspath(__file__))
```

Comment:
what is the value of __file__ at this point wouldn't this code be executing from the agent's egg?

Reply:
`__file__` points to the current file being executed (in this case the `persist_firewall_rules.py` file). I did a manual test using an egg and it maintains the directory structure

Eg: 
>/var/lib/waagent/WALinuxAgent-2.2.48.1/bin/agent/test.egg/azurelinuxagent/agent.py

I needed the path to the current directory to successfully copy over the new network-setup.py file to the correct `bin` directory of the repo (until atleast the daemon is updated and it ensures that all VMs come up with the new file already present there)

Reply:
Does shutil.copyfile reach into the egg?

Reply:
Changed the approach, this is not relevant anymore